lib/remap-lang/src/type_def.rs

```diff
@@ -0,0 +1,127 @@
+use crate::value;
+
+/// Properties for a given expression that express the expected outcome of the
+/// expression.
+///
+/// This includes whether the expression is fallible, whether it can return
+/// "nothing", and a list of values the expression can resolve to.
+#[derive(Debug, Clone, Eq, PartialEq, Default)]
+pub struct TypeDef {
+    /// True, if an expression can return an error.
+    ///
+    /// Some expressions are infallible (e.g. the [`Literal`] expression, or any
+    /// custom function designed to be infallible).
+    pub fallible: bool,
+
+    /// True, if an expression can resolve to "nothing".
+    ///
+    /// For example, and if-statement without an else-condition can resolve to
+    /// nothing if the if-condition does not match.
+    pub optional: bool,
```

Comment:
I don't think `optional` is the best name. It implies that you don't have to use it, which isn't exactly what it means. Perhaps `nullable` would be better?

Reply:
> It implies that you don't have to use it, which isn't exactly what it means.

Could you expand on this a bit more? That is how it is intended to work (and should work) right now.

Having said that, I'm not too worried about this for now, I'm already working on a follow-up PR to rip "optional expressions" out completely, and just have those situations resolve to `null`.

E.g.:

```rust
.foo = if false { "bar" } // .foo == null
```

Reply:
`optional` means you don't have to do something. I think it this case, I guess the implication is that the type doesn't have to resolve to something? However, it is resolving to something - that something being `null`. So I think `nullable` is a more accurate term.

Reply:
That’s incorrect. It _does_ resolve to “nothing”. Because the return value of an expression is `Result<Option<Value>>`.

So `optional` means it can return `Ok(None)`, which is different from `Ok(Some(Null))`.

Having said that. I’ve already pushed a branch that has expressions return `Result<Value>`, and that does simplify things. So we can ignore this for now.

Reply:
Pushed the relevant PR at #5053. That PR does not yet remove the `optional` flag from `TypeDef`, but that will happen in a follow-up PR.